pkg/kubelet/dockertools/manager.go

```diff
@@ -1586,3 +1595,34 @@ func (dm *DockerManager) SyncPod(pod *api.Pod, runningPod kubecontainer.Pod, pod
 
 	return nil
 }
+
+// verifyNonRoot returns an error if the container or image will run as the root user.
+func (dm *DockerManager) verifyNonRoot(container *api.Container) error {
+	if securitycontext.HasRootRunAsUser(container) {
+		return fmt.Errorf("Container's runAsUser breaks non-root policy.")
+	}
+
+	imgRoot, err := dm.isImageRoot(container.Image)
+	if err != nil {
+		return err
+	}
+	if !securitycontext.HasRunAsUser(container) && imgRoot {
+		return fmt.Errorf("Container has no runAsUser and image will run as root.")
+	}
+
+	return nil
+}
+
+// isImageRoot returns true if the user directive is not set on the image, the user is set to 0
+// or the user is set to root.  If there is an error inspecting the image this method will return
+// false and return the error.
+func (dm *DockerManager) isImageRoot(image string) (bool, error) {
+	img, err := dm.client.InspectImage(image)
+	if err != nil {
+		return false, err
+	}
+	if img.Config == nil {
+		return false, fmt.Errorf("Unable to inspect image %s, nil Config", image)
+	}
+	return img.Config.User == "" || img.Config.User == "0" || img.Config.User == "root", nil
```

Comment:
Yes

On Jul 21, 2015, at 2:43 PM, Paul Weil <notifications@github.com> wrote:

In pkg/kubelet/dockertools/manager.go
<https://github.com/GoogleCloudPlatform/kubernetes/pull/11669#discussion_r35138403>
:

> +
> +	return nil
> +}
> +
> +// isImageRoot returns true if the user directive is not set on the image, the user is set to 0
> +// or the user is set to root.  If there is an error inspecting the image this method will return
> +// false and return the error.
> +func (dm *DockerManager) isImageRoot(image string) (bool, error) {
> +	img, err := dm.client.InspectImage(image)
> +	if err != nil {
> +		return false, err
> +	}
> +	if img.Config == nil {
> +		return false, fmt.Errorf("Unable to inspect image %s, nil Config", image)
> +	}
> +	return img.Config.User == "" || img.Config.User == "0" || img.Config.User == "root", nil

We could disallow any non-numeric USER directive in the Dockerfile. I think
that would solve this?

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/11669/files#r35138403>
.